Add docker/docker-compose support for local development

seanwiseman opened the pull request:
### Changes

1. **Dockerisation**: The application has been dockerised with the help of a `docker-compose.yaml` file. This includes the setup for `postgres` and `localstack` services, along with a `migrate` service to handle database migrations and the main `app` service. Thank you @scottaubrey 

2. **Configuration Files**: The configuration files have been updated. The `.docker/app.cfg` file now contains the configuration for the application, including database and other service details.

3. **README Update**: The `README.md` file has been updated to reflect the changes made in this PR. It now includes instructions on how to set up and run the application via docker-compose, as well as information on the new configuration files and environment variables. Key commands have also been documented to assist local article & metrics ingestion. This should help developers understand how to work with the application in its new containerised form.

**Note: I have not made any code changes, the service is untouched. I have also left the previous installation instructions in the README under the `Legacy non-docker installation`**

I have been able to successfully perform local ingestion using the new setup 🎉 

### Next Steps

I will follow up with an additional PR around some local AWS bits and perform some more ingestion tests.

scottaubrey reviewed (changes requested):
Excellent! Thanks so much for the great README that helped me get something up and running further than I had last time. I imported the IDs and can see (empty) results being returned!

I've requested a few docker-ish things to be tweaked particularly around handling sensitive files in a checkout so we don't accidentally bake some credentials in the image at some point! Otherwise, I'm pretty happy with this as a starting point. Thanks so much for taking it on.

ali-tm-amin reviewed (approved):
Thanks @seanwiseman this is great, it looks clean and neat and I module it for other projects.

de-code commented on README.md:
````diff
@@ -21,37 +21,145 @@ would yield a response similar to:
+
+### Build
+```bash
+docker-compose build
````
This is more a question. On another repo, I started to get issues with `docker-compose`. Whereas `docker compose` worked and seem to be the newer version. Just wondering whether we should move to that.

seanwiseman replied:
I will take a look at this 👍🏻

seanwiseman replied:
Both will work `docker-compose` (external tool) and `docker compose` (Docker sub-command), but I have updated the README to use the sub-command 👍🏻

de-code commented:
This is great, thank you. I am looking into more detail especially if we want to use it as a template...


In some other repos we adopted the use of a Makefile mostly as a command entry point (rather than using its dependency capabilities). Just thinking whether we should adopt that here too.

So for example there could be an entry like this (using a command from your README):

```makefile
fetch-metrics:
	docker-compose exec app bash -c "python src/manage.py ingest_metrics"
```

Benefits include for example shell command completion or getting a list of targets. Also similar named targets could be used across a wide variety of projects.

---

Often we use a `docker-compose.override.yml` to define for example mounts, or port mapping. So for example in development, we would mount directories so that they take effect without rebuilding the image (which especially works well with servers that are configured to auto-reload). Whereas in CI we don't want that and we want to make sure the image has everything it needs.

In the data team repos we often do that by defining the Makefile like this:

```makefile
DOCKER_COMPOSE_DEV = docker-compose
DOCKER_COMPOSE_CI = docker-compose -f docker-compose.yml
DOCKER_COMPOSE = $(DOCKER_COMPOSE_DEV)

...

# used by deveeloper
build:
	$(DOCKER_COMPOSE) up -d

start:
	$(DOCKER_COMPOSE) up -d

test:
	$(DOCKER_COMPOSE) run ...

# ci (can re-use other targets)
ci-build-and-test:
	$(MAKE) DOCKER_COMPOSE="$(DOCKER_COMPOSE_CI)" \
		build test
	
```

Some more recent data team related repos are for example [data-hub-api](https://github.com/elifesciences/data-hub-api) (a bit light on documentation) or [sciety-labs](https://github.com/sciety/sciety-labs).

seanwiseman commented:
> In some other repos we adopted the use of a Makefile mostly as a command entry point (rather than using its dependency capabilities). Just thinking whether we should adopt that here too.

I will add a Makefile 👍🏻

seanwiseman commented:
> Often we use a docker-compose.override.yml to define for example mounts, or port mapping

I will look at this 👍🏻

scottaubrey commented:
> This is great, thank you. I am looking into more detail especially if we want to use it as a template...
> 
> In some other repos we adopted the use of a Makefile mostly as a command entry point (rather than using its dependency capabilities). Just thinking whether we should adopt that here too.
> 
> So for example there could be an entry like this (using a command from your README):
> 
> ```makefile
> fetch-metrics:
> 	docker-compose exec app bash -c "python src/manage.py ingest_metrics"
> ```
> 
> Benefits include for example shell command completion or getting a list of targets. Also similar named targets could be used across a wide variety of projects.
> 
> Often we use a `docker-compose.override.yml` to define for example mounts, or port mapping. So for example in development, we would mount directories so that they take effect without rebuilding the image (which especially works well with servers that are configured to auto-reload). Whereas in CI we don't want that and we want to make sure the image has everything it needs.
> 
> In the data team repos we often do that by defining the Makefile like this:
> 
> ```makefile
> DOCKER_COMPOSE_DEV = docker-compose
> DOCKER_COMPOSE_CI = docker-compose -f docker-compose.yml
> DOCKER_COMPOSE = $(DOCKER_COMPOSE_DEV)
> 
> ...
> 
> # used by deveeloper
> build:
> 	$(DOCKER_COMPOSE) up -d
> 
> start:
> 	$(DOCKER_COMPOSE) up -d
> 
> test:
> 	$(DOCKER_COMPOSE) run ...
> 
> # ci (can re-use other targets)
> ci-build-and-test:
> 	$(MAKE) DOCKER_COMPOSE="$(DOCKER_COMPOSE_CI)" \
> 		build test
> 	
> ```
> 
> Some more recent data team related repos are for example [data-hub-api](https://github.com/elifesciences/data-hub-api) (a bit light on documentation) or [sciety-labs](https://github.com/sciety/sciety-labs).

This isn't the best place to have this discussion, granted, but I thought I should throw in that we probably should not use docker compose to build images anyway. It does not have the latest features for building, in particular the ability to build multi-arch images. That requires either docker buildx, and either `docker buildx build` or `docker buildx bake` (which has some support for docker-compose-style config, but not complete from what I understand). 

It would be preferable we move more projects away from `docker compose build` in CI than towards that pattern.

seanwiseman commented:
I believe I have addressed all the feedback now. I think this is good to go.

I will be following up with some small tweaks if needed after my next round of local ingestion tests, but I think this aids local dev for the time being, certainly enough for me to work on some of the small enhancements.
